storage/rocksdb/ha_rocksdb.cc

```diff
@@ -15076,6 +15114,467 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+      (mode & HA_MRR_USE_DEFAULT_IMPL) ||
+      (buf->buffer_end - buf->buffer < mrr_get_length_per_rec()) ||
+      (active_index != table->s->primary_key && (mode & HA_MRR_SORTED))) {
+    mrr_uses_default_impl = true;
+    res = handler::multi_range_read_init(seq, seq_init_param, n_ranges, mode,
+                                         buf);
+    return res;
+  }
+
+  // Ok, using a non-default MRR implementation, MultiGet-MRR
+
+  mrr_uses_default_impl = false;
+  mrr_n_elements = 0;  // nothing to cleanup, yet.
+  mrr_enabled_keyread = false;
+  mrr_rowid_reader = nullptr;
+
+  mrr_funcs = *seq;
+  mrr_buf = *buf;
+
+  bool is_mrr_assoc = !MY_TEST(mode & HA_MRR_NO_ASSOCIATION);
+  if (is_mrr_assoc)
+    status_var_increment(
+        table->in_use->status_var.ha_multi_range_read_init_count);
+
+  mrr_sorted_mode = (mode & HA_MRR_SORTED) ? true : false;
+
+  if (active_index == table->s->primary_key) {
+    mrr_rowid_reader =
+      new Mrr_pk_scan_rowid_source(this, seq_init_param, n_ranges, mode);
+  } else {
+    auto reader = new Mrr_sec_key_rowid_source(this);
+    reader->init(seq, seq_init_param, n_ranges, mode);
+    mrr_rowid_reader = reader;
+  }
+
+  res = mrr_fill_buffer();
+
+  // note: here, we must NOT return HA_ERR_END_OF_FILE even if we know there
+  // are no matches. We should return 0 here and return HA_ERR_END_OF_FILE
+  // from the first multi_range_read_next() call.
+  if (res == HA_ERR_END_OF_FILE) res = 0;
+
+  return res;
+}
+
+// Return the amount of buffer space that MRR scan requires for each record
+// returned
+uint ha_rocksdb::mrr_get_length_per_rec() {
+  return sizeof(rocksdb::Slice) + sizeof(rocksdb::Status) +
+         sizeof(rocksdb::PinnableSlice) +
+         sizeof(char *) +  // this for KEY_MULTI_RANGE::ptr
+         m_pk_descr->max_storage_fmt_length();
+}
+
+
+template <typename T> void align_ptr(char **p) {
+  if (((size_t)p) % alignof(T)) {
+    *p += alignof(T) - ((size_t)p) % alignof(T);
+  }
+}
+
+/*
+  We've got a buffer in mrr_buf, and in order to call RocksDB's MultiGet, we
+  need to use this space to construct several arrays of the same size N:
+
+    rocksdb::Slice[N]         - lookup keys
+    rocksdb::Status[N]        - return statuses
+    rocksdb::PinnableSlice[N] - return rows (*)
+    char*[N]                  - "ptr" value of KEY_MULTI_RANGE. This tells the
+                                SQL layer which lookup key the returned record
+                                matches with (**)
+    {PK lookup value}[N]      - The rowid (Primary Key) to lookup. The
+                                corresponding rocksdb::Slice object points to
+                                this key.
+
+  (*) The memory for rows is allocated somewhere inside RocksDB, there's no
+      way to make it use the user-supplied buffer.
+  (**) The engine could specify HA_MRR_NO_ASSOCIATION which would mean "we
+      cannot tell which key the returned records match" we don't do this.
```

Comment:
`get_quick_select_for_ref` is called when one has
- ref(const) access on the first table 
- "Using filesort" (w/o "Using temporary).

In this case, the query plan is to get filesort() to read the first table and
sort it according to the sort criteria.  filesort is unable to use ref access
as the input stream, it can either read from a quick select or do a full table
scan. Because of that, get_quick_select_for_ref is used.

Reply:
The code here doesn't really handle `HA_MRR_NO_ASSOCIATION`. We just increment the counter, in the same way like DsMrr_impl::dsmrr_init  does:
```
  is_mrr_assoc= !MY_TEST(mode & HA_MRR_NO_ASSOCIATION);
  if (is_mrr_assoc)
    status_var_increment(table->in_use->status_var.ha_multi_range_read_init_count);
```
This allows to use the counter value as a counter of how many times BKA was used.

Some background:
Initially, HA_MRR_NO_ASSOCIATION was introduced for MRR/NDB . NDB returns records but it cannot tell which lookup key each record is for. 

Now, HA_MRR_NO_ASSOCATION is used as follows:
* BKA code does need the storage engine to tell which lookup key the record is for. It doesn't pass HA_MRR_NO_ASSOCATION. 
* Range access doesn't need this, it just needs the record. It passes HA_MRR_NO_ASSOCATION flag.

So, for an MRR implementation, HA_MRR_NO_ASSOCATION means that it doesn't need to keep track which result record matches which range. For  MyRocks/MRR, we won't need the `mrr_range_ptrs` array.  This allows to save a bit of buffer space (8 bytes per range) ...  but you can only do that when you are using range access, and typically don't have many ranges.

Should I implement HA_MRR_NO_ASSOCATION still?

Reply:
Ok